### ISSUE TITLE
Error when installing to DigitalOcean [FAILED! => {"msg": "The task includes an option with an undefined variable..."]

### ISSUE THREAD

**Issue by iamvishnurajan** [trailofbits/algo#14387]
When doing a fresh install to DigitalOcean (NYC3), I get the following error shortly after the droplet is actually created on DigitalOcean. The install itself then does not complete.

`fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'ip_address'\n\nThe error appears to be in '/Users/Vishnu/algo-master/roles/cloud-digitalocean/tasks/main.yml': line 42, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- set_fact:\n  ^ here\n"}`

Full install log is here:

PLAY [localhost] *************************************************************************************

TASK [Gathering Facts] *******************************************************************************
ok: [localhost]

TASK [Playbook dir stat] *****************************************************************************
ok: [localhost]

TASK [Ensure Ansible is not being run in a world writable directory] *********************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}
[WARNING]: The value '' is not a valid IP address or network, passing this value to ipaddr filter
might result in breaking change in future.

TASK [Ensure the requirements installed] *************************************************************
ok: [localhost]

TASK [Set required ansible version as a fact] ********************************************************
ok: [localhost] => (item=ansible-core==2.12.1)

TASK [Verify Python meets Algo VPN requirements] *****************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Verify Ansible meets Algo VPN requirements] ****************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}
[WARNING]: Found variable using reserved name: no_log

PLAY [Ask user for the input] ************************************************************************

TASK [Gathering Facts] *******************************************************************************
ok: [localhost]
[Cloud prompt]
What provider would you like to use?
    1. DigitalOcean
    2. Amazon Lightsail
    3. Amazon EC2
    4. Microsoft Azure
    5. Google Compute Engine
    6. Hetzner Cloud
    7. Vultr
    8. Scaleway
    9. OpenStack (DreamCompute optimised)
    10. CloudStack (Exoscale optimised)
    11. Linode
    12. Install to existing Ubuntu 18.04 or 20.04 server (for more advanced users)
  
Enter the number of your desired provider
:
1^M
TASK [Cloud prompt] **********************************************************************************
ok: [localhost]

TASK [Set facts based on the input] ******************************************************************
ok: [localhost]
[VPN server name prompt]
Name the vpn server
[algo]
:
AlgoVPN-NYC3-20211219^M
TASK [VPN server name prompt] ************************************************************************
ok: [localhost]
[Cellular On Demand prompt]
Do you want macOS/iOS clients to enable "Connect On Demand" when connected to cellular networks?
[y/N]
:
y^M
TASK [Cellular On Demand prompt] *********************************************************************
ok: [localhost]
[Wi-Fi On Demand prompt]
Do you want macOS/iOS clients to enable "Connect On Demand" when connected to Wi-Fi?
[y/N]
:
y^M
TASK [Wi-Fi On Demand prompt] ************************************************************************
ok: [localhost]
[Trusted Wi-Fi networks prompt]
List the names of any trusted Wi-Fi networks where macOS/iOS clients should not use "Connect On Demand"
(e.g., your home network. Comma-separated value, e.g., HomeNet,OfficeWifi,AlgoWiFi)
:
^M
TASK [Trusted Wi-Fi networks prompt] *****************************************************************
ok: [localhost]
[Retain the PKI prompt]
Do you want to retain the keys (PKI)? (required to add users in the future, but less secure)
[y/N]
:
n^M
TASK [Retain the PKI prompt] *************************************************************************
ok: [localhost]
[DNS adblocking prompt]
Do you want to enable DNS ad blocking on this VPN server?
[y/N]
:
y^M
TASK [DNS adblocking prompt] *************************************************************************
ok: [localhost]
[SSH tunneling prompt]
Do you want each user to have their own account for SSH tunneling?
[y/N]
:
y^M
TASK [SSH tunneling prompt] **************************************************************************
ok: [localhost]

TASK [Set facts based on the input] ******************************************************************
ok: [localhost]

PLAY [Provision the server] **************************************************************************

TASK [Gathering Facts] *******************************************************************************
ok: [localhost]

--> Please include the following block of text when reporting issues:

Algo running on: macOS 11.6.2
ZIP file created: Dec 14 15:52:34 2021
Python 3.9.9
Runtime variables:
    algo_provider "digitalocean"
    algo_ondemand_cellular "True"
    algo_ondemand_wifi "True"
    algo_ondemand_wifi_exclude "X251bGw="
    algo_dns_adblocking "True"
    algo_ssh_tunneling "True"
    wireguard_enabled "True"
    dns_encryption "True"

TASK [Display the invocation environment] ************************************************************
changed: [localhost]

TASK [Install the requirements] **********************************************************************
ok: [localhost]

TASK [Generate the SSH private key] ******************************************************************
ok: [localhost]

TASK [Generate the SSH public key] *******************************************************************
ok: [localhost]

TASK [Copy the private SSH key to /tmp] **************************************************************
ok: [localhost]

TASK [Include a provisioning role] *******************************************************************
[cloud-digitalocean : pause]
Enter your API token. The token must have read and write permissions (https://cloud.digitalocean.com/settings/api/tokens):
 (output is hidden):

TASK [cloud-digitalocean : pause] ********************************************************************
ok: [localhost]

TASK [cloud-digitalocean : Set the token as a fact] **************************************************
ok: [localhost]

TASK [cloud-digitalocean : Get regions] **************************************************************
ok: [localhost]

TASK [cloud-digitalocean : Set facts about the regions] **********************************************
ok: [localhost]

TASK [cloud-digitalocean : Set default region] *******************************************************
ok: [localhost]
[cloud-digitalocean : pause]
What region should the server be located in?
    1. ams2     Amsterdam 2
    2. ams3     Amsterdam 3
    3. blr1     Bangalore 1
    4. fra1     Frankfurt 1
    5. lon1     London 1
    6. nyc1     New York 1
    7. nyc2     New York 2
    8. nyc3     New York 3
    9. sfo1     San Francisco 1
    10. sfo2     San Francisco 2
    11. sfo3     San Francisco 3
    12. sgp1     Singapore 1
    13. tor1     Toronto 1
  
Enter the number of your desired region
[8]
:
8^M
TASK [cloud-digitalocean : pause] ********************************************************************
ok: [localhost]

TASK [cloud-digitalocean : Set additional facts] *****************************************************
ok: [localhost]

TASK [cloud-digitalocean : Upload the SSH key] *******************************************************
ok: [localhost]

TASK [cloud-digitalocean : Creating a droplet...] ****************************************************
changed: [localhost]

TASK [cloud-digitalocean : set_fact] *****************************************************************
fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'ip_address'\n\nThe error appears to be in '/Users/Vishnu/algo-master/roles/cloud-digitalocean/tasks/main.yml': line 42, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- set_fact:\n  ^ here\n"}

TASK [include_tasks] *********************************************************************************
included: /Users/Vishnu/algo-master/playbooks/rescue.yml for localhost

TASK [debug] *****************************************************************************************
ok: [localhost] => {
    "fail_hint": [
        "Sorry, but something went wrong!",
        "Please check the troubleshooting guide.",
        "https://trailofbits.github.io/algo/troubleshooting.html"
    ]
}

TASK [Fail the installation] *************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Failed as requested from task"}

PLAY RECAP *******************************************************************************************
localhost                  : ok=35   changed=2    unreachable=0    failed=1    skipped=2    rescued=1    ignored=0

**Comment by gddrt** [comment]
I received this same error attempting to deploy to SFO3 on DigitalOcean with the latest branch (v1.1-134-g7203f33). The droplet was created in my case too.

**Comment by iamvishnurajan** [comment]
I was able to get the install script to complete by editing the following. I think the output from DigitalOcean's API is now different, and so the IP address isn't where the current script expects it. I am not well versed with Ansible, so if someone can confirm the below is safe/accurate, that would be appreciated.

In `/Users/[insert username]/algo-master/roles/cloud-digitalocean/tasks/main.yml`, I edited:

What is presently there towards the end of the file:
```YAML
- block:
  - name: "Create a Floating IP"
    digital_ocean_floating_ip:
      state: present
      oauth_token: "{{ algo_do_token }}"
      droplet_id: "{{ digital_ocean_droplet.data.droplet.id }}"
    register: digital_ocean_floating_ip

  - name: Set the static ip as a fact
    set_fact:
      cloud_alternative_ingress_ip: "{{ digital_ocean_floating_ip.data.floating_ip.ip }}"
  when: alternative_ingress_ip

- set_fact:
    cloud_instance_ip: "{{ digital_ocean_droplet.data.ip_address }}"
    ansible_ssh_user: algo
    ansible_ssh_port: "{{ ssh_port }}"
    cloudinit: true
```

To:
```YAML
- block:
  - name: "Create a Floating IP"
    digital_ocean_floating_ip:
      state: present
      oauth_token: "{{ algo_do_token }}"
      droplet_id: "{{ digital_ocean_droplet.data.droplet.id }}"
    register: digital_ocean_floating_ip

  - name: Set the static ip as a fact
    set_fact:
      cloud_alternative_ingress_ip: "{{ digital_ocean_floating_ip.data.floating_ip.ip }}"

- set_fact:
    cloud_instance_ip: "{{ digital_ocean_floating_ip.data.floating_ip.ip }}"
    ansible_ssh_user: algo
    ansible_ssh_port: "{{ ssh_port }}"
    cloudinit: true
```

For clarity:

- I removed: `when: alternative_ingress_ip`
- Changed `cloud_instance_ip: "{{ digital_ocean_droplet.data.ip_address }}"` to `cloud_instance_ip: "{{ digital_ocean_floating_ip.data.floating_ip.ip }}"`

**Comment by gddrt** [comment]
This issue does not happen in the previous commit (v1.1-133-gc759d75) -- I could set up successfully without editing any files.

**Comment by iamvishnurajan** [comment]
Ah, I suppose the problem must have something to do with the update to ansible core in the latest commit then. I will probably re-deploy with the v1.1-133-gc759d75 version then.

**Comment by davidemyers** [comment]
Try this patch as a workaround:
```patch
diff --git a/roles/cloud-digitalocean/tasks/main.yml b/roles/cloud-digitalocean/tasks/main.yml
index 2013a22..1e6937f 100644
--- a/roles/cloud-digitalocean/tasks/main.yml
+++ b/roles/cloud-digitalocean/tasks/main.yml
@@ -40,7 +40,7 @@
   when: alternative_ingress_ip
 
 - set_fact:
-    cloud_instance_ip: "{{ digital_ocean_droplet.data.ip_address }}"
+    cloud_instance_ip: "{{ digital_ocean_droplet.data.droplet.networks.v4.0.ip_address }}"
     ansible_ssh_user: algo
     ansible_ssh_port: "{{ ssh_port }}"
     cloudinit: true
```

**Comment by 0reo** [comment]
just tried your patch @davidemyers , and got the following error now complaining about the droplet not being defined

`fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'droplet'\n\nThe error appears to be in '/home/oreo/development/algo/roles/cloud-digitalocean/tasks/main.yml': line 42, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- set_fact:\n  ^ here\n"}
`

**Comment by 0reo** [comment]
@davidemyers i checkout out a new copy of the repo, reapplied the patch, and didn't get the same issue this time ¯\_(ツ)_/¯ not sure what i did wrong the first time but must have been something i did

**Comment by troyjfarrell** [comment]
I encountered the same problem as [the first comment on this issue](https://github.com/trailofbits/algo/issues/14387#issue-1084223648).  This diff fixed the problem for me:
```diff
diff --git a/roles/cloud-digitalocean/tasks/main.yml b/roles/cloud-digitalocean/tasks/main.yml
index 2013a22..70afa4f 100644
--- a/roles/cloud-digitalocean/tasks/main.yml
+++ b/roles/cloud-digitalocean/tasks/main.yml
@@ -40,7 +40,7 @@
   when: alternative_ingress_ip
 
 - set_fact:
-    cloud_instance_ip: "{{ digital_ocean_droplet.data.ip_address }}"
+    cloud_instance_ip: "{{ digital_ocean_droplet.data.networks.v4.0.ip_address }}"
     ansible_ssh_user: algo
     ansible_ssh_port: "{{ ssh_port }}"
     cloudinit: true
```

I used
```yaml
- debug:
    var: digital_ocean_droplet
```
to verify that the object at `digital_ocean_droplet.data` matched [the documentation](https://docs.digitalocean.com/reference/api/api-reference/#operation/create_droplet) and that the IPv4 address is now found at `data.networks.v4.0.ip_address` instead of `data.ip_address`.

```shell
$ .env/bin/python -V
Python 3.8.12
$ .env/bin/pip list
Package      Version
------------ -------
ansible      5.0.1
ansible-core 2.12.1
cffi         1.15.0
cryptography 36.0.1
Jinja2       3.0.3
MarkupSafe   2.0.1
netaddr      0.8.0
packaging    21.3
pip          21.3.1
pycparser    2.21
pyOpenSSL    21.0.0
pyparsing    3.0.6
PyYAML       6.0
resolvelib   0.5.4
segno        1.4.1
setuptools   56.0.0
six          1.16.0
```

**Comment by davidemyers** [comment]
@troyjfarrell Your patch fails for me with:
```
The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'networks'
```
When I run `./algo -vvv` I see `networks` as appearing under `droplet`:
```
changed: [localhost] => {
    "changed": true,
    "data": {
        "droplet": {
            "backup_ids": [],
            "created_at": "2021-12-21T15:34:19Z",
            "disk": 25,
            "features": [
                "droplet_agent",
                "ipv6",
                "private_networking"
            ],
            "id": 278807456,
            "image": {
                "created_at": "2021-10-12T21:57:21Z",
                "description": "Ubuntu 20.04 x86",
                "distribution": "Ubuntu",
                "id": 93525508,
                "min_disk_size": 15,
                "name": "20.04 (LTS) x64",
                "public": true,
                "regions": [
                    "nyc3",
                    "nyc1",
                    "sfo1",
                    "nyc2",
                    "ams2",
                    "sgp1",
                    "lon1",
                    "ams3",
                    "fra1",
                    "tor1",
                    "sfo2",
                    "blr1",
                    "sfo3"
                ],
                "size_gigabytes": 0.6,
                "slug": "ubuntu-20-04-x64",
                "status": "available",
                "tags": [],
                "type": "base"
            },
            "kernel": null,
            "locked": false,
            "memory": 1024,
            "name": "vpn8",
            "networks": {
                "v4": [
                    {
                        "gateway": "64.225.0.1",
                        "ip_address": "64.225.3.55",
                        "netmask": "255.255.240.0",
                        "type": "public"
                    },
                    {
                        "gateway": "",
                        "ip_address": "10.132.0.3",
                        "netmask": "255.255.0.0",
                        "type": "private"
                    }
                ],
                "v6": [
                    {
                        "gateway": "2604:a880:800:10::1",
                        "ip_address": "2604:a880:800:10::b04:6001",
                        "netmask": 64,
                        "type": "public"
                    }
                ]
            },
            "next_backup_window": null,
            "region": {
                "available": true,
                "features": [
                    "backups",
                    "ipv6",
                    "metadata",
                    "install_agent",
                    "storage",
                    "image_transfer"
                ],
                "name": "New York 3",
                "sizes": [
                    "s-1vcpu-1gb",
                    "s-1vcpu-1gb-amd",
                    "s-1vcpu-1gb-intel",
                    "s-1vcpu-2gb",
                    "s-1vcpu-2gb-amd",
                    "s-1vcpu-2gb-intel",
                    "s-2vcpu-2gb",
                    "s-2vcpu-2gb-amd",
                    "s-2vcpu-2gb-intel",
                    "s-2vcpu-4gb",
                    "s-2vcpu-4gb-amd",
                    "s-2vcpu-4gb-intel",
                    "s-4vcpu-8gb",
                    "c-2",
                    "c2-2vcpu-4gb",
                    "s-4vcpu-8gb-amd",
                    "s-4vcpu-8gb-intel",
                    "g-2vcpu-8gb",
                    "gd-2vcpu-8gb",
                    "s-8vcpu-16gb",
                    "m-2vcpu-16gb",
                    "c-4",
                    "c2-4vcpu-8gb",
                    "s-8vcpu-16gb-amd",
                    "s-8vcpu-16gb-intel",
                    "m3-2vcpu-16gb",
                    "g-4vcpu-16gb",
                    "so-2vcpu-16gb",
                    "m6-2vcpu-16gb",
                    "gd-4vcpu-16gb",
                    "so1_5-2vcpu-16gb",
                    "m-4vcpu-32gb",
                    "c-8",
                    "c2-8vcpu-16gb",
                    "m3-4vcpu-32gb",
                    "g-8vcpu-32gb",
                    "so-4vcpu-32gb",
                    "m6-4vcpu-32gb",
                    "gd-8vcpu-32gb",
                    "so1_5-4vcpu-32gb",
                    "m-8vcpu-64gb",
                    "c-16",
                    "c2-16vcpu-32gb",
                    "m3-8vcpu-64gb",
                    "g-16vcpu-64gb",
                    "so-8vcpu-64gb",
                    "m6-8vcpu-64gb",
                    "gd-16vcpu-64gb",
                    "so1_5-8vcpu-64gb",
                    "m-16vcpu-128gb",
                    "c-32",
                    "c2-32vcpu-64gb",
                    "m3-16vcpu-128gb",
                    "m-24vcpu-192gb",
                    "g-32vcpu-128gb",
                    "so-16vcpu-128gb",
                    "m6-16vcpu-128gb",
                    "gd-32vcpu-128gb",
                    "m3-24vcpu-192gb",
                    "g-40vcpu-160gb",
                    "so1_5-16vcpu-128gb",
                    "m-32vcpu-256gb",
                    "gd-40vcpu-160gb",
                    "so-24vcpu-192gb",
                    "m6-24vcpu-192gb",
                    "m3-32vcpu-256gb",
                    "m6-32vcpu-256gb"
                ],
                "slug": "nyc3"
            },
            "size": {
                "available": true,
                "description": "Basic",
                "disk": 25,
                "memory": 1024,
                "price_hourly": 0.00744,
                "price_monthly": 5.0,
                "regions": [
                    "ams3",
                    "blr1",
                    "fra1",
                    "lon1",
                    "nyc1",
                    "nyc3",
                    "sfo3",
                    "sgp1",
                    "tor1"
                ],
                "slug": "s-1vcpu-1gb",
                "transfer": 1.0,
                "vcpus": 1
            },
            "size_slug": "s-1vcpu-1gb",
            "snapshot_ids": [],
            "status": "active",
            "tags": [
                "Environment:Algo"
            ],
            "vcpus": 1,
            "volume_ids": [],
            "vpc_uuid": "99496e84-dc84-11e8-80bc-3cfdfea9fba1"
        }
    },
    "invocation": {
        "module_args": {
            "backups": false,
            "firewall": null,
            "id": null,
            "image": "ubuntu-20-04-x64",
            "ipv6": true,
            "monitoring": false,
            "name": "vpn8",
            "private_networking": false,
            "project_name": "",
            "region": "nyc3",
            "resize_disk": false,
            "size": "s-1vcpu-1gb",
            "ssh_keys": [
                "32720199"
            ],
            "tags": [
                "Environment:Algo"
            ],
            "timeout": 30,
            "user_data": "#cloud-config\noutput: {all: '| tee -a /var/log/cloud-init-output.log'}\n\npackage_update: true\npackage_upgrade: true\n\npackages:\n - sudo\n\nusers:\n  - default\n  - name: algo\n    homedir: /home/algo\n    sudo: ALL=(ALL) NOPASSWD:ALL\n    groups: adm,netdev\n    shell: /bin/bash\n    lock_passwd: true\n    ssh_authorized_keys:\n      - \"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC//d8oSfPc64/Wz5uOEaewv0SvlpshzEYIJaSA2slGyaPRAl1ub0LnrqgXBYMLFNdS2CCe2AIUaEXrOqXpBqLqV5WO3u2E7pfpS5Kq4cVi02kw54pRMh131DMghUeFVfVs5ETcUbO9gMRI1j/KrPEKoYW2kuKERMi/Bj8UJK6pHxPjb1oistVKFxMJ7AkXhNnvWhis+ZtP18wlNb23qFnTFUTHyiGKK/8TNW8znr+GHPP3Eix4gyN5zXG/eGa4D3OVCljpAexgisQNvoaR9q0GzZWhDxkOg7gN/Jo2mcuVNXduys9SQqI1/Pmb+hAawTQhGM2J2aHROkmTGRXe/0kx\"\n\nwrite_files:\n  - path: /etc/ssh/sshd_config\n    content: |\n      Port 4160\n      AllowGroups algo\n      PermitRootLogin no\n      PasswordAuthentication no\n      ChallengeResponseAuthentication no\n      UsePAM yes\n      X11Forwarding yes\n      PrintMotd no\n      AcceptEnv LANG LC_*\n      Subsystem\tsftp\t/usr/lib/openssh/sftp-server\n\n\nruncmd:\n  - set -x\n  - sudo apt-get remove -y --purge sshguard || true\n  - systemctl restart sshd.service\n",
            "validate_certs": true,
            "volumes": null,
            "vpc_uuid": null
        }
    }
}
```

**Comment by nkt** [comment]
It seems to Digital Ocean changed API. This replacement fixed issue for me:
```diff
 - set_fact:
-    cloud_instance_ip: "{{ digital_ocean_droplet.data.ip_address }}"
+    cloud_instance_ip: "{{ digital_ocean_droplet.data.networks.v4.0.ip_address }}"
```

**Comment by davidemyers** [comment]
It looks like for some of us DigitalOcean returns `networks` under `droplet` and for some of us it doesn't. That's weird.

**Comment by luixo** [comment]
For those who never used Algo before and are here just for **quick fix**:
Apparently, DigitalOcean API changed.

You can log your `./algo` run with `./algo -vvv`, on the last step you'll be able to see response for DigitalOcean `createDroplet` handler (right after `cloud-digitalocean : Creating a droplet...`). It looks like this (only data needed left):
```
TASK [cloud-digitalocean : Creating a droplet...] *****************************************************************************************
task path: <algo-path>/algo/roles/cloud-digitalocean/tasks/main.yml:12
changed: [localhost] => {
    "changed": true,
    "data": {
        "droplet": {
            "id": 279123910,
            ...
            "networks": {
                "v4": [
                    {
                        "gateway": "157.245.64.1",
                        "ip_address": "157.245.67.6",
                        "netmask": "255.255.240.0",
                        "type": "public"
                    },
                    ...
                ],
                ...
            },
            ...
        }
    },
    ...
}
```

You need to calculate JSON path to `ip_address`, in case above it is `data.droplet.networks.v4.0.ip_address`.
Substitute `data.ip_address` part in `cloud_instance_ip` prop in `algo/roles/cloud-digitalocean/tasks/main.yml` file with the path you evaluated.

In my case I changed `data.ip_address` to `data.droplet.networks.v4.0.ip_address`, but according to other reports you may have a different one.

```diff
 - set_fact:
-    cloud_instance_ip: "{{ digital_ocean_droplet.data.ip_address }}"
+    cloud_instance_ip: "{{ digital_ocean_droplet.data.networks.v4.0.ip_address }}"
```

After that you just need to re-run `./algo` (don't forget to remove your droplet in DigitalOcean console).

**Comment by davidemyers** [comment]
I submitted the patch I suggested above as PR #14390 and it has been merged. If you've had problems deploying to DigitalOcean, download a new copy of Algo and try again.

**Comment by jackivanov** [comment]
Thanks @davidemyers!

**Comment by ifedorenko63** [comment]
I encountered the following problem on a103d8d yesterday:
```
TASK [Wait until SSH becomes ready...] ********************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "elapsed": 321, "msg": "Timeout when waiting for search string OpenSSH in 10.106.0.2:4160"}
```
As you can see, `cloud_instance_ip` was set to private address `10.106.0.2`. This happened because my droplet was created with default private VPC and had 2 addresses - private and public one, and `digital_ocean_droplet.data.droplet.networks.v4.0.ip_address` choses first address, which in my case was the wrong one.

The following diff solved the issue:
```diff
diff --git a/roles/cloud-digitalocean/tasks/main.yml b/roles/cloud-digitalocean/tasks/main.yml
index 1e6937f..780d910 100644
--- a/roles/cloud-digitalocean/tasks/main.yml
+++ b/roles/cloud-digitalocean/tasks/main.yml
@@ -40,7 +40,7 @@
   when: alternative_ingress_ip

 - set_fact:
-    cloud_instance_ip: "{{ digital_ocean_droplet.data.droplet.networks.v4.0.ip_address }}"
+    cloud_instance_ip: "{{ digital_ocean_droplet.data.droplet.networks.v4.1.ip_address }}"
     ansible_ssh_user: algo
     ansible_ssh_port: "{{ ssh_port }}"
     cloudinit: true
```

However, this solution does not really scale as it is not clear how to control the order of addresses DigitalOcean API provides.
The same issue happened in different DigitalOcean regions (ams3, lon1).

**Comment by davidemyers** [comment]
@ifedorenko63 It looks like it's not safe to assume the public address is the first in the list. This is beyond my limited Ansible abilities so I'm going to open a new issue.

**Comment by jackivanov** [comment]
The fix is in #14421

**Comment by sellersshrug0y** [comment]
> I'm still having this issue on [a103d8d](https://github.com/trailofbits/algo/commit/a103d8dd169beddb812863c9d77d5d7ce96d84e4). Perhaps it's just a coincidence, but I've noticed it's only a problem when I select a region that falls outside of my local timezone. In other words, since my local timezone is EST/EDT, I can select the NYC location and everything works great. However, if I select somewhere else (I've tried Amsterdam, London, Frankfort and SF) a droplet is created but the rest of the process fails shortly thereafter with a "Timeout when waiting for search string OpenSSH" error

this fixed my issue: https://github.com/trailofbits/algo/issues/1683#issuecomment-1038196365
